src/core/thread/lowpan.cpp

```diff
@@ -1206,58 +1207,115 @@ int Lowpan::Decompress(Message &           aMessage,
     return (error == OT_ERROR_NONE) ? static_cast<int>(compressedLength) : -1;
 }
 
-otError MeshHeader::Init(const uint8_t *aFrame, uint16_t aFrameLength)
+//---------------------------------------------------------------------------------------------------------------------
+// MeshHeader
+
+void MeshHeader::Init(uint16_t aSource, uint16_t aDestination, uint8_t aHopsLeft)
 {
-    otError error = OT_ERROR_NONE;
+    mSource      = aSource;
+    mDestination = aDestination;
+    mHopsLeft    = aHopsLeft;
+}
+
+bool MeshHeader::IsMeshHeader(const uint8_t *aFrame, uint16_t aFrameLength)
+{
+    return (aFrameLength >= kMinHeaderLength) && ((*aFrame & kDispatchMask) == kDispatch);
+}
+
+otError MeshHeader::ParseFrom(const uint8_t *aFrame, uint16_t aFrameLength, uint16_t &aHeaderLength)
+{
+    otError error = OT_ERROR_PARSE;
+    uint8_t dispatch;
+
+    VerifyOrExit(aFrameLength >= kMinHeaderLength);
+    dispatch = *aFrame++;
 
-    VerifyOrExit(aFrameLength >= 1, error = OT_ERROR_PARSE);
-    mDispatchHopsLeft = *aFrame++;
-    aFrameLength--;
+    VerifyOrExit((dispatch & (kDispatchMask | kSourceShort | kDestShort)) == (kDispatch | kSourceShort | kDestShort));
 
-    if (IsDeepHopsLeftField())
+    mHopsLeft = (dispatch & kHopsLeftMask);
+
+    if (mHopsLeft == kDeepHopsLeft)
     {
-        VerifyOrExit(aFrameLength >= 1, error = OT_ERROR_PARSE);
```

Comment:
Where do we check that the frame length is `>=` than expected header size in the case that Deep Hops Left field is used?

Reply:
Good point. Thanks. Will fix it.